srml/contracts/src/rent.rs

```diff
@@ -95,14 +95,26 @@ fn try_evict_or_and_pay_rent<T: Trait>(
 	// The minimal amount of funds required for a contract not to be evicted.
 	let subsistence_threshold = T::Currency::minimum_balance() + <Module<T>>::tombstone_deposit();
 
+	if balance < subsistence_threshold {
+		// The contract cannot afford to leave a tombstone, so remove the contract info altogether.
+		<ContractInfoOf<T>>::remove(account);
+		runtime_io::kill_child_storage(&contract.trie_id);
+		return RentOutcome::Evicted;
+	}
+
 	let dues = fee_per_block
 		.checked_mul(&blocks_passed.saturated_into::<u32>().into())
 		.unwrap_or(<BalanceOf<T>>::max_value());
-
-	let dues_limited = dues.min(contract.rent_allowance);
-	let rent_allowance_exceeded = dues > contract.rent_allowance;
-	let is_below_subsistence = balance < subsistence_threshold;
-	let go_below_subsistence = balance.saturating_sub(dues_limited) < subsistence_threshold;
+	let rent_budget = contract.rent_allowance.min(balance - subsistence_threshold);
+	let go_below_subsistence = rent_budget < dues;
+
+	// If the rent payment cannot be withdrawn due to locks on the account balance, then evict the
+	// account.
+	//
+	// NOTE: This seems problematic because it provides a way to tombstone an account while
+	// avoiding the last rent payment. In effect, someone could retroactively set rent_allowance
+	// for their contract to 0.
```

Comment:
what does `In effect, someone could retroactively set rent_allowance for their contract to 0.` means ? to change its rent_allowance a contract has to be called thus has to pay its rent

Reply:
Right. So if there's some contract that will be evicted on the next touch and someone wants to prevent it from losing its remaining funds, they would want to set the `rent_allowance` to 0. But, as you note, they can't because that would touch the contract and force a rent payment.

So instead, they use the balances module to place some locks on the funds, which does not evict the contract but has the desired effect since the proposed code will not deduct this locked balance.

Reply:
hmm ok, I understand now, although lock is not easy to call, you only have access indirectly to it though other modules, this is true and great comment